File: packages/apple/Sources/Runtime/Session.swift
import Foundation

/// A stateful AI session that manages conversation state, memory, and guardrails.
///
/// ```swift
/// let session = Session(
///     memory: BufferMemory(),
///     guardrails: [InputLengthGuardrail()]
/// )
/// let response = try await session.send("Hello!")
/// ```
@available(iOS 15.0, macOS 14.0, *)
public final class Session: @unchecked Sendable {
    public let id: String
    private let model: any LocanaraModel
    private let memory: any Memory
    private let guardrails: [any Guardrail]

    public init(
        model: (any LocanaraModel)? = nil,
        memory: (any Memory)? = nil,
        guardrails: [any Guardrail] = []
    ) {
        self.id = UUID().uuidString
        self.model = model ?? LocanaraDefaults.model
        self.memory = memory ?? BufferMemory()
        self.guardrails = guardrails
    }

    /// Send a message and get a response, maintaining conversation state
    public func send(_ message: String) async throws -> String {
        // Apply input guardrails
        var processedText = message
        for guardrail in guardrails {
            switch try await guardrail.checkInput(ChainInput(text: processedText)) {
            case .passed: continue
            case .blocked(let reason): throw LocanaraError.invalidInput("Blocked: \(reason)")
            case .modified(let newText, _): processedText = newText
            }
        }

        // Load memory and build prompt
        let input = ChainInput(text: processedText)
        let entries = await memory.load(for: input)
        var prompt = ""
        for entry in entries {
            prompt += "\(entry.role): \(entry.content)\n"
        }
        prompt += "User: \(processedText)\nAssistant:"

        let response = try await model.generate(prompt: prompt, config: .conversational)

        // Apply output guardrails
        var outputText = response.text
        let output = ChainOutput(value: outputText, text: outputText)
        for guardrail in guardrails {
            switch try await guardrail.checkOutput(output) {
            case .passed: continue
            case .blocked(let reason): throw LocanaraError.executionFailed("Blocked: \(reason)")
            case .modified(let newText, _): outputText = newText
            }
        }

        // Save to memory
        await memory.save(
            input: input,
            output: ChainOutput(value: outputText, text: outputText)
        )

        return outputText
    }

    /// Run a built-in chain within this session context
    public func run(_ chain: any Chain, input: String) async throws -> ChainOutput {
        try await chain.invoke(ChainInput(text: input))
    }

    /// Clear session memory
    public func reset() async {
        await memory.clear()
    }
}

File: packages/android/locanara/src/main/kotlin/com/locanara/runtime/Session.kt
package com.locanara.runtime

import com.locanara.composable.BufferMemory
import com.locanara.composable.Chain
import com.locanara.composable.Guardrail
import com.locanara.composable.GuardrailResult
import com.locanara.composable.Memory
import com.locanara.core.ChainInput
import com.locanara.core.ChainOutput
import com.locanara.core.GenerationConfig
import com.locanara.core.LocanaraDefaults
import com.locanara.core.LocanaraModel
import java.util.UUID

/**
 * A stateful AI session that manages conversation state, memory, and guardrails.
 *
 * ```kotlin
 * val session = Session(
 *     memory = BufferMemory(),
 *     guardrails = listOf(InputLengthGuardrail())
 * )
 * val response = session.send("Hello!")
 * ```
 */
class Session(
    private val model: LocanaraModel = LocanaraDefaults.model,
    private val memory: Memory = BufferMemory(),
    private val guardrails: List<Guardrail> = emptyList()
) {
    val id: String = UUID.randomUUID().toString()

    /** Send a message and get a response, maintaining conversation state */
    suspend fun send(message: String): String {
        var processedText = message
        for (guardrail in guardrails) {
            when (val result = guardrail.checkInput(ChainInput(processedText))) {
                is GuardrailResult.Passed -> continue
                is GuardrailResult.Blocked ->
                    throw IllegalArgumentException("Blocked: ${result.reason}")
                is GuardrailResult.Modified -> processedText = result.newText
            }
        }

        val input = ChainInput(processedText)
        val entries = memory.load(input)
        val prompt = buildString {
            entries.forEach { appendLine("${it.role}: ${it.content}") }
            appendLine("User: $processedText")
            append("Assistant:")
        }

        val response = model.generate(prompt, GenerationConfig.CONVERSATIONAL)

        var outputText = response.text
        val output = ChainOutput(outputText, outputText)
        for (guardrail in guardrails) {
            when (val result = guardrail.checkOutput(output)) {
                is GuardrailResult.Passed -> continue
                is GuardrailResult.Blocked ->
                    throw IllegalStateException("Blocked: ${result.reason}")
                is GuardrailResult.Modified -> outputText = result.newText
            }
        }

        memory.save(input, ChainOutput(outputText, outputText))
        return outputText
    }

    /** Run a built-in chain within this session context */
    suspend fun run(chain: Chain, input: String): ChainOutput {
        return chain.invoke(ChainInput(input))
    }

    /** Clear session memory */
    suspend fun reset() {
        memory.clear()
    }
}
